File: POH7/SantaFuku/Main.cs
// Try POH
// author: Leonardone @ NEETSDKASU
using System;
using System.Collections.Generic;
using System.Linq;

public class Solver
{
    public static void Main()
    {
       int x, y, z, n; Ib(Gis(), out x, out y, out z, out n);
       var xs = Arr(0, x).ToList();
       var ys = Arr(0, y).ToList();
       foreach (var da in NGis(n))
       {
           (da[0] == 0 ? xs : ys).Add(da[1]);
       }
       xs.Sort();
       ys.Sort();
       x = xs.Zip(xs.Skip(1), (a, b) => b - a).Min();
       y = ys.Zip(ys.Skip(1), (a, b) => b - a).Min();
       Console.WriteLine(x * y * z);
    }
    
    ///////////////////////////////////////////////////////////////////
    static string   Gs()  { return Console.ReadLine(); }
    static int      Gi()  { return int.Parse(Gs()); }
    static long     Gl()  { return long.Parse(Gs()); }
    static string[] Gss() { return Gs().Split(" ".ToCharArray()); }
    static int[]    Gis() { return Gss().Select( int.Parse ).ToArray(); }
    static T[]      NGt<T>(int n, Func<T> f) { return Enumerable.Range(0, n).Select( _ => f() ).ToArray(); }
    static string[] NGs(int n) { return NGt(n, Gs); }
    static int[]    NGi(int n) { return NGt(n, Gi); }
    static string[][] NGss(int n) { return NGt(n, Gss); }
    static int[][]    NGis(int n) { return NGt(n, Gis); }
    static void Ib(int[] a, out int v1, out int v2) { v1 = a[0]; v2 = a[1]; }
    static void Ib(int[] a, out int v1, out int v2, out int v3) { v1 = a[0]; v2 = a[1]; v3 = a[2]; }
    static void Ib(int[] a, out int v1, out int v2, out int v3, out int v4) { v1 = a[0]; v2 = a[1]; v3 = a[2]; v4 = a[3]; }
    static T[] Arr<T>(params T[] p) { return p; }
    ///////////////////////////////////////////////////////////////////
}

File: POH7/SantaFuku/Main.vb
' Try POH
' author: Leonardone @ NEETSDKASU
Imports System
Imports Con = System.Console
Imports System.Collections
Imports System.Collections.Generic

Module Main
    
    Sub Main()
        Dim x%, y%, z%, n%, i%, da%()
        Ib(Gis(), x, y, z, n)
        Dim xs As New List(Of Integer)(Arr(0, x))
        Dim ys As New List(Of Integer)(Arr(0, y))
        For Each da In NGis(n)
            If da(0) = 0 Then
                xs.Add(da(1))
            Else
                ys.Add(da(1))
            End If
        Next da
        xs.Sort()
        ys.Sort()
        For i = 1 To xs.Count - 1
            x = Math.Min(x, xs(i) - xs(i - 1))
        Next i
        For i = 1 To ys.Count - 1
            y = Math.Min(y, ys(i) - ys(i - 1))
        Next i
        Con.WriteLine(x * y * z)
    End Sub
    
    ' ----------------------------------------------------------------------------------
    Delegate Function MyFunc(Of T)() As T ' System.Func1(Of T)が何故か使えない
    Function Gs$(): Gs = Console.ReadLine(): End Function
    Function Gi%(): Gi = Int32.Parse(Gs()):  End Function
    Function Gl&(): Gl = Int64.Parse(Gs()):  End Function
    Function Gss() As String(): Gss = Gs().Split(" ".ToCharArray()): End Function
    Function Gis() As Integer(): Dim i%, a%(): Dim s$() = Gss(): ReDim a(s.Length - 1): For i = 0 To s.Length - 1: a(i) = Int32.Parse(s(i)): Next i: Gis = a: End Function
    Function NGt(Of T)(n%, f As MyFunc(Of T)) As T(): Dim a() As T, i%: ReDim a(n - 1): For i = 0 To n - 1: a(i) = f.Invoke(): Next i: NGt = a: End Function
    Function NGs(n%) As String(): NGs = NGt(Of String)(n, AddressOf Gs): End Function
    Function NGi(n%) As Integer(): NGi = NGt(Of Integer)(n, AddressOf Gi): End Function
    Function NGss(n%) As String()(): NGss = NGt(Of String())(n, AddressOf Gss): End Function
    Function NGis(n%) As Integer()(): NGis = NGt(Of Integer())(n, AddressOf Gis): End Function
    Function SubArr(a() As Integer, s%) As Integer(): Dim r() As Integer: If s < a.Length Then: ReDim r(a.Length - s - 1): Array.Copy(a, s, r, 0, r.Length): End If: SubArr = r: End Function
    Function SubArr(a() As Integer, s%, l%) As Integer(): Dim r() As Integer: If l > 0 Then: ReDim r(l - 1): Array.Copy(a, s, r, 0, l): End If: SubArr = r: End Function
    Function Ib(a() As Integer, ByRef v1%) As Integer(): v1 = a(0): Ib = SubArr(a, 1): End Function
    Function Ib(a() As Integer, ByRef v1%, ByRef v2%) As Integer(): v1 = a(0): v2 = a(1): Ib = SubArr(a, 2): End Function
    Function Ib(a() As Integer, ByRef v1%, ByRef v2%, ByRef v3%) As Integer(): v1 = a(0): v2 = a(1): v3 = a(2): Ib = SubArr(a, 3): End Function
    Function Ib(a() As Integer, ByRef v1%, ByRef v2%, ByRef v3%, ByRef v4%) As Integer(): v1 = a(0): v2 = a(1): v3 = a(2): v4 = a(3): Ib = SubArr(a, 4): End Function
    Function Arr(ByVal ParamArray a() As Integer) As Integer(): Arr = a: End Function
    ' ----------------------------------------------------------------------------------
End Module
